common/types/string.go

```diff
@@ -221,3 +240,447 @@ func stringStartsWith(s String, pre ref.Val) ref.Val {
 	}
 	return Bool(strings.HasPrefix(string(s), string(preStr)))
 }
+
+func stringFormat(s String, args ref.Val) ref.Val {
+	return stringFormatWithLocale(s, args, String(defaultLocale))
+}
+
+func stringFormatWithLocale(s String, args ref.Val, localeVal ref.Val) ref.Val {
+	formatStr := string(s)
+	argList, ok := args.(traits.Lister)
+	if !ok {
+		return MaybeNoSuchOverloadErr(argList)
+	}
+	localeStrVal, ok := localeVal.(String)
+	if !ok {
+		return MaybeNoSuchOverloadErr(localeStrVal)
+	}
+	locale := localeStrVal.Value().(string)
+
+	i := 0
+	argIndex := 0
+	var builtStr strings.Builder
+	for i < len(formatStr) {
+		if formatStr[i] == '%' {
+			if i+1 < len(formatStr) && formatStr[i+1] == '%' {
+				err := builtStr.WriteByte('%')
+				if err != nil {
+					return NewErr("error writing format string: %s", err.Error())
+				}
+				i += 2
+				continue
+			} else {
+				newStrIndex, val, refErr := parseFormatArg(i, argIndex, formatStr, argList, locale)
+				if refErr != nil {
+					return refErr
+				}
+				_, err := builtStr.WriteString(val)
+				if err != nil {
+					return NewErr("error writing format string: %s", err.Error())
+				}
+				i = newStrIndex
+				argIndex++
+			}
+		} else {
+			err := builtStr.WriteByte(formatStr[i])
+			if err != nil {
+				return NewErr("error writing format string: %s", err.Error())
+			}
+			i++
+		}
+	}
+	return String(builtStr.String())
+}
+
+func parseFormatArg(lastStrIndex, argIndex int, formatStr string, args traits.Lister, locale string) (int, string, ref.Val) {
+	if lastStrIndex+1 >= len(formatStr) {
+		return -1, "", NewErr("unexpected end of string")
+	}
+	refSize := args.Size()
+	argSize, ok := refSize.(Int)
+	if !ok {
+		return -1, "", NewErr("couldn't convert arg size")
+	}
+	if Int(argIndex) >= argSize {
+		return -1, "", NewErr("index %d out of range", argIndex)
+	}
+
+	i := lastStrIndex
+	i, formatter, err := parseFormattingClause(i, formatStr)
+	if err != nil {
+		return -1, "", NewErr("could not parse formatting clause: %s", err)
+	}
+
+	var argStr string
+	argAny := args.Get(Int(argIndex))
+	if formatter == nil {
+		return -1, "", NewErr("formatter required")
+	} else {
+		argStr, err = formatter(argAny, locale)
+		if err != nil {
+			return -1, "", NewErr("error during formatting: %s", err)
+		}
+	}
+	return i, argStr, nil
+}
+
+type clauseImpl func(ref.Val, string) (string, error)
+
+func formatString(lastStrIndex int, formatStr string) (int, clauseImpl, error) {
+	return lastStrIndex + 1, func(arg ref.Val, locale string) (string, error) {
+		argStrVal := arg.ConvertToType(StringType)
+		argStr, ok := argStrVal.Value().(string)
+		if !ok {
+			return "", fmt.Errorf("could not convert argument %q to string", argStrVal)
+		}
+		return argStr, nil
+	}, nil
+}
+
+func formatFixed(lastStrIndex int, formatStr string, precision *int) (int, clauseImpl, error) {
+	if precision == nil {
```

Comment:
Why not use a default if a precision is not provided?

Reply:
Right, I think for other developers it'd be unusual. I'll set the default to 6 for now, but I can change it if others have thoughts.